text/0000-async_await.md

```diff
@@ -0,0 +1,571 @@
+services space, and we wish to continue to enable those users - and to enable
+adoption by other users - by making it more ergonomic to write asynchronous
+network services in Rust.
+
+The development of asynchronous IO in Rust has gone through multiple phases.
+Prior to 1.0, we experimented with having a green-threading runtime built into
+the language. However, this proved too opinionated - because it impacted every
+program written in Rust - and it was removed shortly before 1.0. After 1.0,
+asynchronous IO initially focused around the mio library, which provided an
+cross-platform abstraction over the async IO primitives of Linux, Mac OS, and
+Windows. In mid-2016, the introduction of the futures crate had a major impact
+by providing a convenient, shared abstraction for asynchronous operations. The
+tokio library provided a mio-based event loop that could execute code
+implemented using the futures interfaces.
+
+After gaining experience & user feedback with the futures-based ecosystem, we
+discovered certain ergonomics challenges. Using state which needs to be shared
+across await points was extremely ergonomic - requiring either Arcs or join
+chaining - and while combinators were often more ergonomic than manually
+writing a future, they still often led to messy sets of nested and chained
+callbacks.
+
+Fortunately, the Future abstraction is well suited to use with a syntactic
+sugar which has become common in many languages with async IO - the async and
+await keywords. In brief, an asynchronous function returns a future, rather
+than evaluating immediately when it is called. Inside the function, other
+futures can be awaited using an await expression, which causes them to yield
+control while the future is being polled. From a user's perspective, they can
+use async/await as if it were synchronous code, and only need to annotate their
+functions and calls.
+
+Async/await & futures can be a powerful abstraction for asynchronicity and
+concurrency in general, and likely has applications outside of the asynchronous
+IO space. The use cases we've experience with today are generally tied to async
+IO, but by introducing first class syntax and libstd support we believe more
+use cases for async & await will also flourish, that are not tied directly to
+asynchronous IO.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Async functions
+
+Functions can be annotated with the `async` keyword, making them "async
+functions":
+
+```rust
+async fn function(argument: &str) -> usize {
+     // ...
+}
+```
+
+Async functions work differently from normal functions. When an async function
+is called, it does not enter the body immediately. Instead, it evaluates to an
+anonymous type which implements `Future`. As that future is polled, the
+function is evaluated up to the next `await` or return point inside of it (see
+the await syntax section next).
```

Comment:
The dart language also chose to have `async` functions not enter the body immediately, but then chose to change to evaluating up until the first `await` at `2.0` of the language. There reasons for doing so are documented [here](https://github.com/dart-lang/sdk/blob/master/docs/newsletter/20170915.md#synchronous-async-start). It might be worth evaluating moving to that model.

Reply:
There has been a lot of discussion of that very issue in this very thread. The eager evaluation model does not work well for Rust, but there are ways to mitigate the downsides.

(See @aturon's last summary comment for more specific references.)